pandas/tseries/offsets.py

```diff
@@ -76,7 +76,22 @@
 # DateOffset
 
 
-class DateOffset(BaseOffset):
+class OffsetMeta(type):
+    """
+    Metaclass that allows us to pretend that all BaseOffset subclasses
+    inherit from DateOffset (which is needed for backward-compatibility).
+    """
+
+    @classmethod
+    def __instancecheck__(cls, obj) -> bool:
+        return isinstance(obj, BaseOffset)
+
+    @classmethod
+    def __subclasscheck__(cls, obj) -> bool:
```

Comment:
@simonjayhawkins is this going to mess up annotations elsewhere we annotate something as a DateOffset?

Reply:
mypy is a static checker, so IIUC mypy doesn't execute the code in \_\_instancecheck__ and \_\_subclasscheck__. Therefore, mypy has no way of knowing that BaseOffset subclass pretend to be DateOffset subclasses. 

This is probably a use case for Protocol. i.e. structural typing instead of nominal typing.